[CoSimulationApplication] 3D domain (volume) - 1D domain (line) data transfer operator

loumalouomega opened the pull request:
**📝 Description**

This is a very "not processional" PR, I don't like the quality of the result, and I was doing this in a private development, but due to certain requirements in @KratosMultiphysics/altair, I am uploading it to the main repository.

This adds  `DataTransfer3D1DProcess` and the corresponding *CoSimumlation data transfer operator*, in order to transfer from a 1D domain to a 3D domain, working similarly to a mapper. In the 3D-1D works the same as a mapper. In order to do so, it creates an auxiliary intersected modelpart between the 3D and 1D domain, then uses the mapper.

Originally was purely tetra-line, but I replaced the ` IntersectionUtilities` from #10463 with the `IsInside` from geometries. This is more generic and robust, but in some corner cases (line intersect only one corner) could give a bit worse results. In the future, I will create a meta-mapper like the 2D-3D, which will replace current implementation, as this requires extra work for MPI, and we need it now I do it simply as a data transfer operator.

Includes #10463 (despite is not needed any more). 

The tests included:

- **Block:**

![image](https://user-images.githubusercontent.com/19991680/211400498-346a8059-1449-4be0-a80e-5f815f960c7b.png)

- **Torus:**

![image](https://user-images.githubusercontent.com/19991680/211399544-27ebdb05-69fe-486f-bfa1-b35477d2c821.png)

**🆕 Changelog**

- Added #10463
- Added `DataTransfer3D1DProcess`
- Added tests

loumalouomega committed:
WIP

loumalouomega committed:
More detailed info

loumalouomega committed:
A lot of WIP

loumalouomega committed:
Improvement

loumalouomega committed:
Adding check

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
WIP

loumalouomega committed:
Fix

loumalouomega committed:
WIP

loumalouomega committed:
WIP

loumalouomega committed:
[Core] Adding EraseNonHistoricalVariable to VariableUtils

loumalouomega committed:
Expose to python

loumalouomega committed:
Add test

loumalouomega committed:
Suggestion

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Unused

loumalouomega committed:
More WIP

loumalouomega committed:
Even more WIP

loumalouomega committed:
Reduce code duplication

loumalouomega committed:
Normalize

loumalouomega committed:
Missing

loumalouomega committed:
AtomicUtilities missing

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Name consistency

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Corrections

loumalouomega committed:
Missing

loumalouomega committed:
Minor

loumalouomega committed:
[Core] Consistent definition between InvertMatrix and GeneralizedInvertMatrix

loumalouomega committed:
Merge branch 'core/intersection-line-tetrahedra' into cosim/tetra-line-data-transfer

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Refactoring

loumalouomega committed:
Reduce code duplication

loumalouomega committed:
More code reduction

loumalouomega committed:
WIP

loumalouomega committed:
Setting flags

loumalouomega committed:
More refactor

loumalouomega committed:
Missing check

loumalouomega committed:
More refactor

loumalouomega committed:
Missing

loumalouomega committed:
Missing

loumalouomega committed:
Mssing

loumalouomega committed:
Minor

loumalouomega committed:
Minor check

loumalouomega committed:
Minor

loumalouomega committed:
Simplification

loumalouomega committed:
Improve mapping settings

loumalouomega committed:
Increase even more

loumalouomega committed:
WIP

loumalouomega committed:
More WIP

loumalouomega committed:
Missing

loumalouomega committed:
Update

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Some suggestions

loumalouomega committed:
Using enums as suggested

loumalouomega committed:
Update

loumalouomega committed:
Refactor code to improve readibility

loumalouomega committed:
Merge master

loumalouomega committed:
Refactor (I)

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Refactor all

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
[Core] Missing Flags derivation from Process when exposing to python

loumalouomega committed:
Better flag

loumalouomega committed:
Cleaning

loumalouomega committed:
Changes

loumalouomega committed:
Minor

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Cleaning

loumalouomega committed:
Adding tests

loumalouomega committed:
Minor

loumalouomega committed:
Minor clean up

loumalouomega committed:
Unused variables

loumalouomega committed:
Codacy suggestion

philbucher reviewed (changes requested):
Before reviewing this in detail, please clean it up a bit

The size of the test needs to be reduced (best to make a new clean pr so that this does not end up in the enormous git-history)

Also please separate the core changes

loumalouomega committed:
Cleaning

loumalouomega committed:
Minor

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Update

loumalouomega committed:
Consistency

loumalouomega committed:
Minor

loumalouomega committed:
Minor update

loumalouomega committed:
Minor

loumalouomega committed:
Memory fix

loumalouomega committed:
Fixes

loumalouomega committed:
Clear is evil

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Reduce mesh sizes

loumalouomega committed:
Revert

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Update

philbucher commented on applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py:
```diff
@@ -0,0 +1,256 @@
+import os
+
+def GetFilePath(fileName):
+    return os.path.join(os.path.dirname(os.path.realpath(__file__)), fileName)
```
`pathlib.Path`

philbucher commented on applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py:
```diff
@@ -0,0 +1,256 @@
+        # # Debug
+        debug_vtk(self.current_model, ["Block", "Line"])
+
+    @KratosUnittest.skipIf(KM.IsDistributedRun(), "This test is designed for serial runs only.")
```
```suggestion
```

those are not needed, we never use this in other tests

philbucher commented on applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py:
```diff
@@ -0,0 +1,256 @@
+    def tearDown(self):
+        pass
+
+    @KratosUnittest.skipIf(KM.IsDistributedRun(), "This test is designed for serial runs only.")
```
```suggestion
```

philbucher commented on applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py:
```diff
@@ -0,0 +1,256 @@
+        # # Debug
+        debug_vtk(self.current_model, ["Torus", "Circle"])
+
+    @KratosUnittest.skipIf(KM.IsDistributedRun(), "This test is designed for serial runs only.")
```
```suggestion
```

philbucher commented on applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py:
```diff
@@ -0,0 +1,256 @@
+    def tearDown(self):
+        pass
+
+    @KratosUnittest.skipIf(KM.IsDistributedRun(), "This test is designed for serial runs only.")
```
```suggestion
```

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp:
```diff
@@ -0,0 +1,290 @@
+{
+    auto& r_elements_array = rModelPart.Elements();
+    KRATOS_ERROR_IF(r_elements_array.size() == 0) << "Empty model part" << std::endl;
+    return block_for_each<MaxReduction<double>>(r_elements_array, [&](Element& rElement) {
```
```suggestion
    return block_for_each<MaxReduction<double>>(r_elements_array, [&](const auto& rElement) {
```

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h:
```diff
@@ -0,0 +1,306 @@
+     * @param rModelPart The model part to compute
+     * @return The maximum length
+     */
+    static double GetMaxLength(ModelPart& rModelPart);
```
```suggestion
    static double GetMaxLength(const ModelPart& rModelPart);
```

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp:
```diff
@@ -0,0 +1,290 @@
+
+double DataTransfer3D1DProcess::GetMaxLength(ModelPart& rModelPart)
+{
+    auto& r_elements_array = rModelPart.Elements();
```
what about conditions?

Conditions have priority over elements in the mappingApp

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp:
```diff
@@ -0,0 +1,290 @@
+    const auto it_elem_begin = r_elements_array.begin();
+
+    /// KDtree definitions
+    KRATOS_KDTREE_POINTELEMENT_DEFINITION
```
what does this do?

loumalouomega replied:
It is a macro defined in the header

ChengshunShang1996 committed:
add a measurement for density with periodic condition

ChengshunShang1996 committed:
first servo loading - not check yet

ChengshunShang1996 committed:
a stable servo loading

ChengshunShang1996 committed:
fix the test case of moving boundaries

philbucher commented on applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py:
```diff
@@ -0,0 +1,256 @@
+        vtk_output_parameters = KratosMultiphysics.Parameters("""{
+            "Parameters" : {
+                "model_part_name"                    : "",
+                "file_format"                        : "ascii",
```
```suggestion
```

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py:
```diff
@@ -0,0 +1,139 @@
+    def _Check(self, from_solver_data, to_solver_data):
+        def CheckData(data_to_check):
+            if "node" not in data_to_check.location:
+                raise Exception('Transfer only supports nodal values!"{}"\nChecking ModelPart "{}" of solver "{}"'.format(self._ClassName(), data_to_check.model_part_name, data_to_check.solver_name))
```
please use f-strings in new code

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py:
```diff
@@ -0,0 +1,139 @@
+        CheckData(from_solver_data)
+        CheckData(to_solver_data)
+
+    def __check_model_part_3D(self, model_part):
```
what if the ModelPart has 1D and 3D elements?

loumalouomega replied:
I will raise an exception

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py:
```diff
@@ -0,0 +1,139 @@
+            if self.__check_model_part_3D(model_part_origin):
+                self.origin_is_3d = True
+            else:
+                self.origin_is_3d = False
```
```suggestion
            self.origin_is_3d = self.__check_model_part_3D(model_part_origin)
```

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.cpp:
```diff
@@ -0,0 +1,290 @@
+
+DataTransfer3D1DProcess::~DataTransfer3D1DProcess()
+{
+}
```
```suggestion
```

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h:
```diff
@@ -0,0 +1,306 @@
+    }
+
+    /// Destructor.
+    ~PointElement() override= default;
```
```suggestion
```

unused code => remove

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h:
```diff
@@ -0,0 +1,306 @@
+        }
+
+        // Second model part
+        if (rSecondModelPart.NumberOfElements() > 0 || rSecondModelPart.NumberOfConditions() > 0 ) {
```
conditions take precedence over elements in the mappingApp FYI

philbucher reviewed (commented):
can you please reduce the size of the mdpa files?

ChengshunShang1996 committed:
add test case for servo control

loumalouomega committed:
Remove unused

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Remove unused

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Remove unused

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Remove unused

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer

loumalouomega committed:
Refactor file path handling in test_3d_1d_data_transfer_process.py

loumalouomega committed:
Refactor test_3d_1d_data_transfer_process.py

loumalouomega committed:
Reduce mdpa sizes

loumalouomega committed:
Suggestion

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Unused

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
More unused

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Refactor data_transfer_3D_1D_process.h

loumalouomega committed:
Refactor data_transfer_3D_1D_process.cpp and data_transfer_3D_1D_process.h

loumalouomega committed:
Suggestion

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
More suggestion

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Suggestion

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Refactor GetMaxLength function in data_transfer_3D_1D_process.cpp

loumalouomega committed:
Refactor error message in Kratos3D1DDataTransferOperator class

loumalouomega committed:
Refactor initialization and execution of data transfer processes

This commit refactors the initialization and execution of data transfer processes in the `Kratos3D1DDataTransferOperator` class. The `__init__` method now initializes the data transfer object with the given settings and sets up internal structures for managing data transfer processes between solvers. The `_ExecuteTransferData` method now executes the data transfer between two solvers, either by retrieving and using an existing data transfer process or creating a new one if none exists. The `_Check` method validates that the source and destination data contain only nodal data for transfer operations. The `__check_model_part_3D` method checks whether the given model part consists entirely of 3D elements and verifies that all elements have the same geometry type.

philbucher commented on applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py:
```diff
@@ -0,0 +1,279 @@
+    }""")
+
+    out_parameters["model_part_name"].SetString(domain)
+    out_parameters["output_file_name"].SetString(str(output_filename) + "_data_transfer.json")
```
```suggestion
    out_parameters["output_file_name"].SetString(f"{output_filename}_data_transfer.json")
```

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py:
```diff
@@ -0,0 +1,276 @@
+        if first_geometry_type == 1:
+            return False
+        else:
+            return True
```
```suggestion
        return first_geometry_type != 1
```

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py:
```diff
@@ -0,0 +1,276 @@
+        """
+        first_geometry_type = None
+
+        for elem in model_part.Elements:
```
again, what about conditions?

loumalouomega replied:
True

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h:
```diff
@@ -0,0 +1,297 @@
+        );
+
+    /// Destructor.
+    ~DataTransfer3D1DProcess() override = default;
```
```suggestion
```

loumalouomega replied:
Without this does not compile

philbucher commented on applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h:
```diff
@@ -0,0 +1,297 @@
+ * @param rSecondModelPart The second ModelPart
+ * @return The 1D model part 
+ */
+ModelPart& Determine1DModelPart(ModelPart& rFirstModelPart, ModelPart& rSecondModelPart)
```
can you move the logic to the source file?

loumalouomega replied:
Okay

markelov208 committed:
[GeoMechanicsApplication] Making unit tests for the erosion mechanism (#12789)

* first version of unit tests

* corrected second unit test

* removed unused definition

* formatted

* formatetd 2.0

* added this-> to call of BaseClassFinalizeSolutionStep

* updated README file

* changed GravitationalForce

* changed to GravitationalAcceleration

* response to all reviews

* fixed that

* response to review on geo_mechanics_newton_raphson_erosion_process_strategy

* removed some include, p_scheme and p_solver

philbucher reviewed (dismissed):
LGTM, left minor comments

loumalouomega committed:
Update applications/CoSimulationApplication/tests/test_3d_1d_data_transfer_process.py

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Update applications/CoSimulationApplication/python_scripts/data_transfer_operators/3d_1d_data_transfer.py

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Update applications/CoSimulationApplication/custom_processes/data_transfer_3D_1D_process.h

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Check both conditions and elements

loumalouomega committed:
Move methods to source file

loumalouomega committed:
Merge branch 'master' into cosim/tetra-line-data-transfer
